src/main/java/Request/RequestParser.java

```diff
@@ -1,30 +1,73 @@
+import java.io.BufferedReader;
+import java.io.IOException;
+
 public class RequestParser {
-  private String method;
-  private RequestLine requestLine;
-  private String[] splitRequest;
-  private String uri;
-  private String version;
-
-  public RequestParser(String requestString) {
-    this.splitRequest = requestString.split("\r\n");
+  BufferedReader reader; 
+
+  public RequestParser(BufferedReader reader) {
+    this.reader = reader;
   }
-  
+
   public Request generateRequest() throws BadRequestException {
-    parseRequestLine();
-    return new Request.Builder()
-                      .method(this.requestLine.getMethod())
-                      .uri(this.requestLine.getURI())
-                      .version(this.requestLine.getHTTPVersion())
-                      .build();
+    try {
+      Request.Builder requestBuilder = new Request.Builder();
+      buildRequestLine(requestBuilder);
+      buildHeaders(requestBuilder);
+      buildMessageBody(requestBuilder);
+      return requestBuilder.build();
+    } catch (ArrayIndexOutOfBoundsException e) {
+      throw new BadRequestException("Bad request from ArrayIndex!");
+    } catch (IOException e) {
+      throw new BadRequestException("Bad request from IOException!");
+    }
   }
 
-  private void parseRequestLine() throws BadRequestException {
-    try {
-      String requestLineString = this.splitRequest[0];
-      this.requestLine = new RequestLine(requestLineString);
-    } catch (Exception e) {
-      throw new BadRequestException("Bad request!");
+  private Request.Builder buildRequestLine(Request.Builder requestBuilder) throws IOException {
```

Comment:
You aren't the return values of any of these private methods here. Why not just return each value and build the request in `generateRequest`?

```java
RequestLine requestLine = new RequestLine(this.read.readLine());
HashMap<String, String> headers = parseHeaders();
HashMap<String, String> body = parseMessageBody();

return new Request.Builder()
  .method(requestLine.getMethod())
  .uri(requestLine.getURI())
  .version(requestLine.getHTTPVersion())
  .headers(headers)
  .body(body)
  .build();
```

Reply:
I've followed this suggestion, which actually allowed me to improve my exception handling. Instead of catching `ArrayIndexOutOfBoundException` and `IOException` inside the `#generateRequest` method, I know catch those exceptions inside every individual helper method. Upon catching them, I throw a `BadRequestException` whose message points to the part of the incoming request that caused the exception.  This approach allows `ClientThread` to catch the `BadRequestException` and build the correct `Response` for the end user while simultaneously providing useful information to clients of the class.